File: sandbox_terminal/utils/sandbox_manager.py
"""
Core sandbox management logic that coordinates all components.
"""

import asyncio
import os
import shutil
import logging
from pathlib import Path
from typing import Dict, List, Optional, Any, Tuple
from datetime import datetime

from .session_store import SessionStore, SessionMetadata
from .docker_sandbox import DockerSandbox, DOCKER_AVAILABLE
from .file_tracker import FileChangeTracker

logger = logging.getLogger("sandbox_terminal.sandbox_manager")


class SandboxManager:
    """Manages sandbox sessions and coordinates between components."""
    
    def __init__(self, storage_path: Optional[str] = None):
        self.session_store = SessionStore(storage_path)
        self.docker_sandbox = DockerSandbox() if DOCKER_AVAILABLE else None
        self.file_trackers: Dict[str, FileChangeTracker] = {}
        
        # Configuration
        self.max_sandboxes = int(os.environ.get("MAX_SANDBOXES", "10"))
        self.max_sandbox_size = int(os.environ.get("MAX_SANDBOX_SIZE", "1000")) * 1024 * 1024  # MB to bytes
        
        # Clean up any orphaned containers on startup
        if self.docker_sandbox:
            self._cleanup_orphaned_containers()
            
    def create_sandbox(self, source_path: str, session_name: str, exclude: Optional[List[str]] = None) -> Dict[str, Any]:
        """Create a new sandbox session."""
        # Validate inputs
        source_path = Path(source_path).resolve()
        if not source_path.exists():
            raise ValueError(f"Source path does not exist: {source_path}")
        if not source_path.is_dir():
            raise ValueError(f"Source path must be a directory: {source_path}")
            
        # Check session name
        if self.session_store.get_session(session_name):
            raise ValueError(f"Session '{session_name}' already exists")
            
        # Check sandbox limit
        active_sessions = [s for s in self.session_store.list_sessions() if s.status == "active"]
        if len(active_sessions) >= self.max_sandboxes:
            raise ValueError(f"Maximum number of sandboxes ({self.max_sandboxes}) reached")
            
        try:
            # Create session
            session = self.session_store.create_session(session_name, str(source_path))
            session.exclude_patterns = exclude
            self.session_store.update_session(session)
            
            # Copy files to sandbox
            logger.info(f"Copying files from {source_path} to {session.sandbox_path}")
            if exclude:
                logger.info(f"Excluding patterns: {exclude}")
            self._copy_directory(source_path, session.sandbox_path, exclude)
            
            # Check size limit
            sandbox_size = self._get_directory_size(session.sandbox_path)
            if sandbox_size > self.max_sandbox_size:
                # Clean up and fail
                self.session_store.delete_session(session_name)
                shutil.rmtree(session.sandbox_path, ignore_errors=True)
                raise ValueError(f"Sandbox size ({sandbox_size / 1024 / 1024:.1f}MB) exceeds limit ({self.max_sandbox_size / 1024 / 1024}MB)")
                
            # Initialize file tracker
            tracker = FileChangeTracker(session.sandbox_path)
            tracker.capture_initial_state()
            self.file_trackers[session_name] = tracker
            
            # Save initial snapshot
            snapshot_path = Path(session.sandbox_path).parent / "initial_snapshot.json"
            tracker.save_snapshot(str(snapshot_path))
            
            # Create Docker container if available
            if self.docker_sandbox:
                try:
                    container = self.docker_sandbox.create_container(
                        session_name,
                        session.sandbox_path,
                        session.environment
                    )
                    session.container_id = container.id
                    self.docker_sandbox.start_container(container)
                    self.session_store.update_session(session)
                except Exception as e:
                    logger.error(f"Failed to create Docker container: {e}")
                    # Continue without Docker
                    
            return {
                "success": True,
                "session_name": session_name,
                "source_path": str(source_path),
                "sandbox_path": session.sandbox_path,
                "size": f"{sandbox_size / 1024 / 1024:.1f}MB",
                "container_id": session.container_id,
                "created_at": session.created_at
            }
            
        except Exception as e:
            # Clean up on failure
            self.session_store.delete_session(session_name)
            if session_name in self.file_trackers:
                del self.file_trackers[session_name]
            raise
            
    def destroy_sandbox(self, session_name: str) -> Dict[str, Any]:
        """Destroy a sandbox session and clean up resources."""
        session = self.session_store.get_session(session_name)
        if not session:
            raise ValueError(f"Session '{session_name}' not found")
            
        # Stop and remove Docker container
        if self.docker_sandbox and session.container_id:
            try:
                container = self.docker_sandbox.get_container(session_name)
                if container:
                    self.docker_sandbox.stop_container(container)
                    self.docker_sandbox.remove_container(container)
            except Exception as e:
                logger.error(f"Failed to remove container: {e}")
                
        # Remove file tracker
        if session_name in self.file_trackers:
            del self.file_trackers[session_name]
            
        # Delete session (this also removes files)
        self.session_store.delete_session(session_name)
        
        return {
            "success": True,
            "message": f"Sandbox '{session_name}' destroyed successfully"
        }
        
    def list_sandboxes(self) -> List[Dict[str, Any]]:
        """List all sandbox sessions."""
        sessions = self.session_store.list_sessions()
        
        result = []
        for session in sessions:
            # Get size
            size = self.session_store.get_session_size(session.session_id)
            
            # Get container status
            container_status = "not_created"
            if self.docker_sandbox and session.container_id:
                container = self.docker_sandbox.get_container(session.session_id)
                if container:
                    container_status = container.status
                    
            result.append({
                "name": session.session_id,
                "source": session.source_path,
                "created": session.created_at,
                "last_accessed": session.last_accessed,
                "size": f"{size / 1024 / 1024:.1f}MB",
                "status": session.status,
                "container_status": container_status,
                "command_count": len(session.command_history)
            })
            
        return sorted(result, key=lambda x: x["last_accessed"], reverse=True)
        
    def execute_command(
        self,
        session_name: str,
        command: str,
        working_dir: Optional[str] = None,
        timeout: int = 30
    ) -> Dict[str, Any]:
        """Execute a command in a sandbox."""
        session = self.session_store.get_session(session_name)
        if not session:
            raise ValueError(f"Session '{session_name}' not found")
            
        # Validate working directory
        if working_dir:
            # Ensure it's relative and within sandbox
            if os.path.isabs(working_dir):
                working_dir = working_dir.lstrip("/")
            work_path = Path(session.sandbox_path) / working_dir
            if not work_path.exists():
                raise ValueError(f"Working directory does not exist: {working_dir}")
                
        # Update session working directory
        if working_dir:
            session.working_dir = working_dir
            self.session_store.update_session(session)
            
        # Execute command
        start_time = datetime.now()
        
        if self.docker_sandbox and session.container_id:
            # Execute in Docker container
            container = self.docker_sandbox.get_container(session_name)
            if not container:
                raise RuntimeError(f"Container not found for session '{session_name}'")
                
            output, exit_code = self.docker_sandbox.execute_command(
                container,
                command,
                working_dir=f"/workspace/{working_dir}" if working_dir else None,
                environment=session.environment,
                timeout=timeout
            )
        else:
            # Fallback to subprocess execution
            import subprocess
            
            # Build safe environment
            env = os.environ.copy()
            env.update(session.environment)
            env["HOME"] = session.sandbox_path
            env["PWD"] = str(Path(session.sandbox_path) / (working_dir or ""))
            
            try:
                proc = subprocess.run(
                    command,
                    shell=True,
                    cwd=Path(session.sandbox_path) / (working_dir or ""),
                    env=env,
                    capture_output=True,
                    text=True,
                    timeout=timeout
                )
                output = proc.stdout + proc.stderr
                exit_code = proc.returncode
            except subprocess.TimeoutExpired:
                output = "Command timed out"
                exit_code = -1
            except Exception as e:
                output = str(e)
                exit_code = -1
                
        execution_time = (datetime.now() - start_time).total_seconds()
        
        # Log to history
        self.session_store.add_command_to_history(
            session_name,
            command,
            output,
            exit_code,
            execution_time
        )
        
        return {
            "success": exit_code == 0,
            "session_name": session_name,
            "command": command,
            "working_dir": working_dir or session.working_dir,
            "output": output,
            "exit_code": exit_code,
            "execution_time": f"{execution_time:.2f}s"
        }
        
    async def execute_command_async(
        self,
        session_name: str,
        command: str,
        working_dir: Optional[str] = None,
        timeout: int = 30
    ) -> Dict[str, Any]:
        """Execute a command in a sandbox asynchronously."""
        # For now, just run the sync version in an executor
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            self.execute_command,
            session_name,
            command,
            working_dir,
            timeout
        )
        
    async def create_sandbox_async(
        self,
        source_path: str,
        session_name: str,
        exclude: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """Create a new sandbox session asynchronously."""
        # Run the sync version in an executor to avoid blocking
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            self.create_sandbox,
            source_path,
            session_name,
            exclude
        )
        
    def get_file_changes(self, session_name: str) -> Dict[str, Any]:
        """Get file changes in a sandbox."""
        session = self.session_store.get_session(session_name)
        if not session:
            raise ValueError(f"Session '{session_name}' not found")
            
        tracker = self.file_trackers.get(session_name)
        if not tracker:
            # Recreate tracker if needed
            tracker = FileChangeTracker(session.sandbox_path)
            snapshot_path = Path(session.sandbox_path).parent / "initial_snapshot.json"
            if snapshot_path.exists():
                tracker.load_snapshot(str(snapshot_path))
            else:
                tracker.capture_initial_state()
            self.file_trackers[session_name] = tracker
            
        changes = tracker.get_changes()
        summary = tracker.get_diff_summary()
        
        return {
            "success": True,
            "session_name": session_name,
            "changes": changes,
            "summary": summary
        }
        
    def reset_sandbox(self, session_name: str) -> Dict[str, Any]:
        """Reset a sandbox to its original state."""
        session = self.session_store.get_session(session_name)
        if not session:
            raise ValueError(f"Session '{session_name}' not found")
            
        # Remove current workspace
        shutil.rmtree(session.sandbox_path, ignore_errors=True)
        
        # Copy original files again
        source_path = Path(session.source_path)
        # Use stored exclude patterns from session
        self._copy_directory(source_path, session.sandbox_path, session.exclude_patterns)
        
        # Reset file tracker
        if session_name in self.file_trackers:
            tracker = self.file_trackers[session_name]
            tracker.reset_to_original()
            
        # Clear command history
        session.command_history = []
        session.working_dir = "/"
        self.session_store.update_session(session)
        
        return {
            "success": True,
            "message": f"Sandbox '{session_name}' reset to original state"
        }
        
    def _copy_directory(self, src: Path, dst: str, exclude: Optional[List[str]] = None) -> None:
        """Copy a directory to sandbox location."""
        dst_path = Path(dst)
        
        # Remove destination if it exists
        if dst_path.exists():
            shutil.rmtree(dst_path)
            
        # Ensure parent directory exists
        dst_path.parent.mkdir(parents=True, exist_ok=True)
        
        # Create ignore function if exclude patterns provided
        ignore_func = None
        if exclude:
            import fnmatch
            def ignore_patterns(directory, filenames):
                """Function to determine which files to ignore during copy."""
                ignored = set()
                
                # Calculate relative path of current directory from source
                try:
                    current_rel_path = Path(directory).relative_to(src)
                except ValueError:
                    current_rel_path = Path()
                
                for pattern in exclude:
                    # Handle directory patterns (ending with /)
                    if pattern.endswith('/'):
                        dir_pattern = pattern.rstrip('/')
                        # Check directory names
                        for name in filenames:
                            full_path = Path(directory) / name
                            if full_path.is_dir() and fnmatch.fnmatch(name, dir_pattern):
                                ignored.add(name)
                    else:
                        # Handle both files and directories for other patterns
                        for name in filenames:
                            full_path = Path(directory) / name
                            
                            # Build relative path from source
                            if current_rel_path == Path():
                                item_rel_path = name
                            else:
                                item_rel_path = str(current_rel_path / name)
                            
                            # Check against pattern
                            if '**' in pattern:
                                # Handle recursive patterns
                                # For directories, check if pattern matches directory structure
                                if full_path.is_dir():
                                    # Check if this directory or its contents would match
                                    # For example, **/__pycache__ should match any __pycache__ dir
                                    pattern_parts = pattern.split('/')
                                    if pattern_parts[0] == '**' and len(pattern_parts) > 1:
                                        # Pattern like **/__pycache__
                                        target_name = pattern_parts[1]
                                        if fnmatch.fnmatch(name, target_name):
                                            ignored.add(name)
                                else:
                                    # For files, use full relative path matching
                                    if fnmatch.fnmatch(item_rel_path, pattern):
                                        ignored.add(name)
                            else:
                                # Simple pattern matching
                                if fnmatch.fnmatch(name, pattern):
                                    ignored.add(name)
                                    
                return list(ignored)
            ignore_func = ignore_patterns
            
        # Use shutil for the copy
        shutil.copytree(src, dst, symlinks=True, ignore_dangling_symlinks=True, ignore=ignore_func)
        
    def _get_directory_size(self, path: str) -> int:
        """Get total size of a directory."""
        total_size = 0
        for dirpath, dirnames, filenames in os.walk(path):
            for filename in filenames:
                filepath = os.path.join(dirpath, filename)
                if os.path.exists(filepath):
                    total_size += os.path.getsize(filepath)
        return total_size
        
    def _cleanup_orphaned_containers(self) -> None:
        """Clean up any orphaned Docker containers."""
        if not self.docker_sandbox:
            return
            
        try:
            # Get all sandbox containers
            containers = self.docker_sandbox.list_sandbox_containers()
            
            # Get all active sessions
            active_sessions = {s.session_id for s in self.session_store.list_sessions()}
            
            # Remove containers for non-existent sessions
            for container in containers:
                session_name = container.name.replace("mcp-sandbox-", "")
                if session_name not in active_sessions:
                    logger.info(f"Removing orphaned container: {container.name}")
                    try:
                        container.remove(force=True)
                    except:
                        pass
                        
        except Exception as e:
            logger.error(f"Failed to cleanup orphaned containers: {e}")

File: node_sandbox_terminal/src/utils/sandbox-manager.ts
import { promises as fs } from "fs";
import path from "path";
import { v4 as uuidv4 } from "uuid";
import { glob } from "glob";
import { minimatch } from "minimatch";
import * as fse from "fs-extra";
import {
  SandboxSession,
  SandboxConfig,
  CommandResult,
  SandboxDiff,
  FileChange,
} from "../types/index.js";
import { DockerSandbox } from "./docker-sandbox.js";
import { CommandValidator } from "./command-validator.js";
import { FileTracker } from "./file-tracker.js";
import { SessionStore } from "./session-store.js";

export class SandboxManager {
  private config: SandboxConfig;
  private docker: DockerSandbox;
  private validator: CommandValidator;
  private sessionStore: SessionStore;

  constructor(config: SandboxConfig) {
    this.config = config;
    this.docker = new DockerSandbox(config);
    this.validator = new CommandValidator(config.security);
    this.sessionStore = new SessionStore(path.join(config.storagePath, "sessions"));
  }

  async createSandbox(
    sourcePath: string,
    sessionName: string,
    allowedPaths?: string[],
    excludePatterns?: string[]
  ): Promise<SandboxSession> {
    // Validate session name
    if (!/^[a-zA-Z0-9_-]+$/.test(sessionName)) {
      throw new Error("Session name must contain only letters, numbers, hyphens, and underscores");
    }

    // Check if session already exists
    const existing = await this.sessionStore.getSession(sessionName);
    if (existing) {
      throw new Error(`Session '${sessionName}' already exists`);
    }

    // Check concurrent limit
    const sessions = await this.sessionStore.listSessions();
    if (sessions.length >= this.config.maxConcurrent) {
      throw new Error(`Maximum concurrent sandboxes (${this.config.maxConcurrent}) reached`);
    }

    // Create sandbox directory
    const sandboxId = uuidv4();
    const sandboxPath = path.join(this.config.storagePath, "sandboxes", sandboxId);
    await fse.ensureDir(sandboxPath);

    try {
      // Copy source directory to sandbox
      await this.copyDirectory(sourcePath, sandboxPath, excludePatterns);

      // Create session
      const session: SandboxSession = {
        id: sandboxId,
        name: sessionName,
        source: sourcePath,
        sandboxPath,
        created: new Date(),
        lastAccessed: new Date(),
        workingDir: "/workspace",
        environment: {},
        commandHistory: [],
        allowedPaths,
        networkEnabled: this.config.security.networkEnabled,
      };

      // Create Docker container
      const containerId = await this.docker.createContainer(session);
      session.containerId = containerId;

      // Save session
      await this.sessionStore.saveSession(session);

      return session;
    } catch (error) {
      // Cleanup on failure
      await fse.remove(sandboxPath);
      throw error;
    }
  }

  async listSandboxes(): Promise<Array<{
    name: string;
    source: string;
    created: string;
    size: string;
    active: boolean;
  }>> {
    const sessions = await this.sessionStore.listSessions();
    const results = [];

    for (const session of sessions) {
      const size = await this.getDirectorySize(session.sandboxPath);
      const active = session.containerId ? await this.docker.isContainerRunning(session.containerId) : false;
      
      results.push({
        name: session.name,
        source: session.source,
        created: session.created.toISOString(),
        size: this.formatBytes(size),
        active,
      });
    }

    return results;
  }

  async executeCommand(
    sessionName: string,
    command: string,
    workingDir?: string,
    timeout?: number
  ): Promise<CommandResult> {
    const session = await this.getSession(sessionName);

    // Validate command
    if (!this.validator.isCommandSafe(command)) {
      throw new Error("Command blocked for security reasons");
    }

    // Update session
    session.lastAccessed = new Date();
    if (workingDir) {
      session.workingDir = path.join("/workspace", workingDir);
    }

    // Execute in container
    const result = await this.docker.executeCommand(
      session,
      command,
      timeout || this.config.defaultTimeout
    );

    // Update command history
    session.commandHistory.push({
      command,
      exitCode: result.exitCode,
      output: result.stdout,
      error: result.stderr,
      timestamp: new Date(),
      workingDir: session.workingDir,
    });

    // Limit history size
    if (session.commandHistory.length > 100) {
      session.commandHistory = session.commandHistory.slice(-100);
    }

    await this.sessionStore.saveSession(session);

    return result;
  }

  async readFile(sessionName: string, filePath: string): Promise<string> {
    const session = await this.getSession(sessionName);
    const fullPath = path.join(session.sandboxPath, filePath);

    // Validate path is within sandbox
    if (!this.isPathInSandbox(fullPath, session.sandboxPath)) {
      throw new Error("Access denied: path outside sandbox");
    }

    return await fs.readFile(fullPath, "utf-8");
  }

  async writeFile(sessionName: string, filePath: string, content: string): Promise<void> {
    const session = await this.getSession(sessionName);
    const fullPath = path.join(session.sandboxPath, filePath);

    // Validate path is within sandbox
    if (!this.isPathInSandbox(fullPath, session.sandboxPath)) {
      throw new Error("Access denied: path outside sandbox");
    }

    // Ensure directory exists
    await fse.ensureDir(path.dirname(fullPath));
    
    // Write file
    await fs.writeFile(fullPath, content);

    // Update session
    session.lastAccessed = new Date();
    await this.sessionStore.saveSession(session);
  }

  async listFiles(sessionName: string, dirPath: string): Promise<Array<{
    name: string;
    type: "file" | "directory";
    size: number;
    modified: string;
  }>> {
    const session = await this.getSession(sessionName);
    const fullPath = path.join(session.sandboxPath, dirPath);

    // Validate path is within sandbox
    if (!this.isPathInSandbox(fullPath, session.sandboxPath)) {
      throw new Error("Access denied: path outside sandbox");
    }

    const entries = await fs.readdir(fullPath, { withFileTypes: true });
    const results = [];

    for (const entry of entries) {
      const entryPath = path.join(fullPath, entry.name);
      const stats = await fs.stat(entryPath);

      results.push({
        name: entry.name,
        type: entry.isDirectory() ? "directory" as const : "file" as const,
        size: stats.size,
        modified: stats.mtime.toISOString(),
      });
    }

    return results.sort((a, b) => {
      if (a.type !== b.type) {
        return a.type === "directory" ? -1 : 1;
      }
      return a.name.localeCompare(b.name);
    });
  }

  async getDiff(sessionName: string): Promise<SandboxDiff> {
    const session = await this.getSession(sessionName);
    const tracker = new FileTracker(session.source, session.sandboxPath);
    return await tracker.calculateDiff();
  }

  async resetSandbox(sessionName: string): Promise<void> {
    const session = await this.getSession(sessionName);

    // Remove sandbox directory
    await fse.remove(session.sandboxPath);

    // Recreate from source
    await fse.ensureDir(session.sandboxPath);
    await this.copyDirectory(session.source, session.sandboxPath);

    // Restart container
    if (session.containerId) {
      await this.docker.removeContainer(session.containerId);
      session.containerId = await this.docker.createContainer(session);
    }

    // Reset session state
    session.commandHistory = [];
    session.workingDir = "/workspace";
    session.lastAccessed = new Date();

    await this.sessionStore.saveSession(session);
  }

  async destroySandbox(sessionName: string): Promise<void> {
    const session = await this.getSession(sessionName);

    // Remove container
    if (session.containerId) {
      await this.docker.removeContainer(session.containerId);
    }

    // Remove sandbox directory
    await fse.remove(session.sandboxPath);

    // Remove session
    await this.sessionStore.deleteSession(sessionName);
  }

  async commitChanges(
    sessionName: string,
    targetPath: string,
    previewOnly = false
  ): Promise<SandboxDiff | { committed: boolean; diff: SandboxDiff }> {
    const session = await this.getSession(sessionName);
    const diff = await this.getDiff(sessionName);

    if (previewOnly) {
      return diff;
    }

    // Apply changes
    for (const file of diff.deleted) {
      const targetFile = path.join(targetPath, file.path);
      if (await fse.pathExists(targetFile)) {
        await fse.remove(targetFile);
      }
    }

    for (const file of [...diff.added, ...diff.modified]) {
      const sourceFile = path.join(session.sandboxPath, file.path);
      const targetFile = path.join(targetPath, file.path);
      await fse.ensureDir(path.dirname(targetFile));
      await fse.copy(sourceFile, targetFile, { overwrite: true });
    }

    return { committed: true, diff };
  }

  private async getSession(sessionName: string): Promise<SandboxSession> {
    const session = await this.sessionStore.getSession(sessionName);
    if (!session) {
      throw new Error(`Session '${sessionName}' not found`);
    }
    return session;
  }

  private async copyDirectory(
    source: string,
    destination: string,
    excludePatterns?: string[]
  ): Promise<void> {
    const files = await glob("**/*", {
      cwd: source,
      dot: true,
      nodir: true,
    });

    for (const file of files) {
      // Check exclude patterns
      if (excludePatterns?.some(pattern => minimatch(file, pattern))) {
        continue;
      }

      const sourcePath = path.join(source, file);
      const destPath = path.join(destination, file);

      await fse.ensureDir(path.dirname(destPath));
      await fse.copy(sourcePath, destPath);
    }
  }

  private isPathInSandbox(testPath: string, sandboxPath: string): boolean {
    const relative = path.relative(sandboxPath, testPath);
    return !relative.startsWith("..") && !path.isAbsolute(relative);
  }

  private async getDirectorySize(dirPath: string): Promise<number> {
    let size = 0;
    const files = await glob("**/*", {
      cwd: dirPath,
      dot: true,
      nodir: true,
    });

    for (const file of files) {
      const filePath = path.join(dirPath, file);
      const stats = await fs.stat(filePath);
      size += stats.size;
    }

    return size;
  }

  private formatBytes(bytes: number): string {
    if (bytes === 0) return "0 B";
    const k = 1024;
    const sizes = ["B", "KB", "MB", "GB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${(bytes / Math.pow(k, i)).toFixed(1)} ${sizes[i]}`;
  }
}